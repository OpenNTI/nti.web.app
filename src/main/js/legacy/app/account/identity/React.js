const Ext = require('@nti/extjs');
const { getService } = require('@nti/web-client');
const { Identity } = require('@nti/web-profiles');
const LoginActions = require('internal/legacy/login/Actions');
require('internal/legacy/overrides/ReactHarness');

const AccountActions = require('../Actions');

module.exports = exports = Ext.define(
	'NextThought.app.account.identity.React',
	{
		extend: 'NextThought.ReactHarness',
		alias: 'widget.identity-react',
		component: Identity.Session,

		setTheme(scope, knockout) {},

		initComponent: function () {
			this.callParent(arguments);
			Object.assign(this.initialConfig, {
				addHistory: true,
				baseroute: '/app',
			});
			this.AccountActions = AccountActions.create();
			this.LoginActions = LoginActions.create();
		},

		async afterRender() {
			this.callParent(arguments);
			const service = await getService();
			const supportLinks = service.getSupportLinks();

			this.mon(Ext.getBody(), {
				click: this.onBodyClicks.bind(this, supportLinks),
			});
		},

		onBodyClicks(supportLinks, e) {
			let action = e.getTarget('a[data-action]')?.dataset.action;
			const link = e.getTarget('a[href]');
			const href = link?.getAttribute('href');

			// This is used to launch support logic the same way everywhere
			// we have a link that goes to the supportContact url. (so we can
			// just emit that url in the dom and not worry about coping logic)
			if (
				supportLinks.supportContact === href &&
				supportLinks.internalSupport
			) {
				action = action || 'support';
			}

			if (action) {
				if (call(this, `do-${action}`, link, e) !== false) {
					e.stopEvent();
				}
			}
		},

		showLink(target) {
			target.setAttribute('target', '_blank');
			return false;
		},

		'do-welcomeGuide'(target) {
			this.AccountActions.showWelcomePage(target.href);
		},

		'do-about': 'showLink',

		'do-privacy'(target) {
			this.AccountActions.showPrivacy(target);
		},

		'do-privacyForMinors'(target) {
			this.AccountActions.showChildrensPrivacy(target.href);
		},

		'do-terms'(target) {
			this.AccountActions.showTermsOfService(target.href);
		},

		'do-support'(target) {
			this.AccountActions.showContactUs();
		},

		'do-helpSite': 'showLink',

		'do-impersonate'() {
			this.LoginActions.handleImpersonate();
		},

		'do-logout'() {
			this.LoginActions.handleLogout();
		},
	}
);

function call(scope, action, ...args) {
	if (typeof action === 'string') {
		return call(scope, scope[action], ...args);
	}

	return action.apply(scope, args);
}
