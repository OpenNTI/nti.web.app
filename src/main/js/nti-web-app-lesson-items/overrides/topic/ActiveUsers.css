.active-users {
	--side-padding: 1rem;

	background: white;
	position: sticky;
	top: -18px;
}

.active-users > :global(.empty-state-component) {
	color: var(--secondary-grey);
	margin: 0;
}

.heading {
	padding: 1.125rem var(--side-padding, 1rem);
	font: normal 600 1rem/1.3125 var(--body-font-family);
}

.user-list {
	list-style: none;
	padding: 0 0 0.5rem;
	margin: 0;
}

.user-list li {
	margin: 0;
}

.user-list-item {
	display: flex;
	flex-direction: row;
	align-items: center;
	padding: 0.5rem var(--side-padding, 1rem);
	border-top: 1px solid var(--border-grey-light);
	overflow: hidden;
}

.user-list-item .avatar {
	flex: 0 0 auto;
	width: 40px;
	height: 40px;
	margin-right: 0.25rem;
}

.user-list-item .display-name {
	flex: 1 1 auto;
	font: normal 600 0.875rem/1.3 var(--body-font-family);
	color: var(--secondary-grey);
	white-space: nowrap;
	overflow: hidden;
	text-overflow: ellipsis;
}

