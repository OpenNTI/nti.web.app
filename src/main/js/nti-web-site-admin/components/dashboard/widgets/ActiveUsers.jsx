import { NumericValue } from '@nti/web-charts';
import { getService } from '@nti/web-client';
import { getLink } from '@nti/lib-interfaces';
import { scoped } from '@nti/lib-locale';
import { useResolver } from '@nti/web-commons';

const SESSIONS = 'Sessions';
const ANALYTICS = 'Analytics';
const ACTIVE_SESSION_COUNT = 'active_session_count';

const t = scoped(
	'nti-web-site-admin.components.dashboard.widgets.ActiveSessions',
	{
		sessionsUnavailable: 'Unable to get active session data',
		activeSessionsLabel: 'Users Online Now',
	}
);

const { isResolved } = useResolver;

export function ActiveUsers() {
	const resolver = useResolver(async () => {
		const service = await getService();

		try {
			const sessionsCollection = service.getCollection(
				SESSIONS,
				ANALYTICS
			);
			const link = getLink(sessionsCollection, ACTIVE_SESSION_COUNT);
			const stats = await service.get(link);

			return { value: stats.Count };
		} catch (e) {
			return {
				value: 0,
				error: <div className="not-available">{t('notAvailable')}</div>,
			};
		}
	});

	return isResolved(resolver) ? (
		<NumericValue label={t('activeSessionsLabel')} value={resolver.value} />
	) : (
		<div />
	);
}
