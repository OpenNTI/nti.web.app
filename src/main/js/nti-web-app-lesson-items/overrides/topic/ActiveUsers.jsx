import React from 'react';
import PropTypes from 'prop-types';
import classnames from 'classnames/bind';
import {LinkTo} from '@nti/web-routing';
import {User, EmptyState, List} from '@nti/web-commons';
import {scoped} from '@nti/lib-locale';

import Styles from './ActiveUsers.css';

const cx = classnames.bind(Styles);
const t = scoped('NTIWebAppLessonItems.overrides.topic.ActiveUsers', {
	heading: 'Active People',
	empty: 'There are no active people.'
});

export default class ActiveUsers extends React.Component {
	static propTypes = {
		activeUsers: PropTypes.array
	}

	render () {
		const {activeUsers} = this.props;

		if (!activeUsers) { return null; }

		const empty = activeUsers.length === 0;

		return (
			<div className={cx('active-users')}>
				{empty && (<EmptyState subHeader={t('empty')} />)}
				{!empty && (<div className={cx('heading')}>{t('heading')}</div>)}
				{!empty && (
					<List.Unadorned className={cx('user-list')}>
						{activeUsers.map((user, index) => {
							return (
								<li key={index}>
									<LinkTo.Object className={cx('user-list-item')} object={user}>
										<User.Avatar className={cx('avatar')} user={user} />
										<User.DisplayName className={cx('display-name')} user={user} />
									</LinkTo.Object>
								</li>
							);
						})}
					</List.Unadorned>
				)}
			</div>
		);
	}
}
