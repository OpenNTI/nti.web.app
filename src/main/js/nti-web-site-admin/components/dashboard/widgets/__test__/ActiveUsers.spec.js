import React from 'react';
import { create, act } from 'react-test-renderer';

import * as TestUtils from '@nti/web-client/test-utils';
import { flushPromises } from '@nti/lib-commons/test-utils';

import { ActiveUsers } from '../ActiveUsers';

const { tearDownTestClient, setupTestClient } = TestUtils;

const getMockService = Count => {
	return {
		getCollection: () => {
			return {
				hasLink: () => true,
				getLink: () => 'mockLink',
			};
		},
		get: () => {
			return Promise.resolve({
				Count,
			});
		},
	};
};

const onBefore = count => {
	jest.useFakeTimers();
	setupTestClient(getMockService(count));
};

const onAfter = () => {
	tearDownTestClient();
};

/* eslint-env jest */
describe('Site admin dashboard widget active sessions (with 5 count)', () => {
	beforeEach(() => onBefore(5));
	afterEach(onAfter);

	test('Basic render test', async () => {
		let cmp;
		await act(async () => {
			cmp = create(<ActiveUsers />);
			jest.runAllTimers();
			await flushPromises();
			jest.runAllTimers();
		});

		const tree = cmp.toJSON();

		expect(tree).toMatchSnapshot();
		cmp.unmount();
	});
});

describe('Site admin dashboard widget active sessions (with 0 count)', () => {
	beforeEach(() => onBefore(0));
	afterEach(onAfter);

	test('Basic render test', async () => {
		let cmp;
		await act(async () => {
			cmp = create(<ActiveUsers />);
			jest.runAllTimers();
			await flushPromises();
			jest.runAllTimers();
		});

		const tree = cmp.toJSON();

		expect(tree).toMatchSnapshot();
		cmp.unmount();
	});
});
