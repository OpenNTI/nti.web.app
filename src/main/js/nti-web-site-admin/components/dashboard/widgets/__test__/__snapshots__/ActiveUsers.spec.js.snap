// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Site admin dashboard widget active sessions (with 0 count) Basic render test 1`] = `
<div
  className="numeric-value"
>
  <div
    className="label"
  >
    Users Online Now
  </div>
  <div
    className="value"
  >
    0
  </div>
</div>
`;

exports[`Site admin dashboard widget active sessions (with 5 count) Basic render test 1`] = `
<div
  className="numeric-value"
>
  <div
    className="label"
  >
    Users Online Now
  </div>
  <div
    className="value"
  >
    5
  </div>
</div>
`;
